todo_app/app.py

```diff
@@ -1,59 +1,138 @@
     app = Flask(__name__)
     app.config.from_object('todo_app.flask_config.Config')
 
     todo_board_id = os.getenv('TODO_BOARD_ID')
+    github_client_id = os.getenv('CLIENT_ID')
+    github_client_secret = os.getenv('CLIENT_SECRET')
+
+    app.config['LOGIN_DISABLED'] = os.getenv('LOGIN_DISABLED') == 'True'
+
+    login_manager = LoginManager()
+    login_manager.anonymous_user = MyCustomAnonymousUser
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        client = WebApplicationClient(github_client_id)
+        return redirect(client.prepare_request_uri('https://github.com/login/oauth/authorize'))
+        
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User(user_id)
 
+    login_manager.init_app(app) 
+
+    @app.route('/login/callback')
+    def login():
+        #TODO - state check
+        
+        request_code = request.args.get('code')
+        client = WebApplicationClient(github_client_id)
+        
+        token_url, headers, body = client.prepare_token_request(
+            'https://github.com/login/oauth/access_token',
+            authorization_response=request.url,
+            code=request_code,
+            client_secret=github_client_secret
+        )
+
+        token_resp = requests.post(token_url, headers=headers, data=body)
+        client.parse_request_body_response(token_resp.text)
+
+        get_user_uri, headers, body = client.add_token('https://api.github.com/user')
+        user_info_resp = requests.get(get_user_uri, headers=headers, data=body)
+        user_id=user_info_resp.json()["id"]
+        user = User(user_id)
+        login_user(user)
+
+        return redirect('/')
 
     @app.route('/')
+    @login_required
     def index():
         todo_board = ToDoBoard(todo_board_id)
         items = todo_board.get_cards()
-        item_view_model = ViewModel(items)
+        item_view_model = ViewModel(items, current_user.role)
         return render_template('index.html', view_model=item_view_model)
 
     
     @app.route('/items', methods=['POST'])
+    @login_required
+    @writer_required
     def add_item():
         title = request.form.get('title')
         status = request.form.get('status')
         todo_board = ToDoBoard(todo_board_id)
         todo_board.create_card(title, status)
 
         items = todo_board.get_cards()
-        item_view_model = ViewModel(items)
+        item_view_model = ViewModel(items, current_user.role)
         return redirect('/')
 
     @app.route('/items/<item_id>')
+    @login_required
+    @writer_required
     def update_item_status(item_id):    
         status = request.args.get('status')
         todo_board = ToDoBoard(todo_board_id)
         todo_board.update_card_status(item_id, status)
 
         items = todo_board.get_cards()
-        item_view_model = ViewModel(items)
+        item_view_model = ViewModel(items, current_user.role)
         return render_template('index.html', view_model=item_view_model)
```

Comment:
As opposed to the above, you do use the ViewModel here but it would be nicer to keep on handing responsibility to the `index` function via a redirect rather than updating three identical sections of code